### PR TITLE
Implement titlecase iterator

### LINKED ISSUE #65: Implement titlecase iterator
Implement a `Titlecase` iterator similar to the `Lowercase` and `Uppercase` iterators already in this crate.

Unlike `Lowercase` and `Uppercase`, `char::to_titlecase` is not defined in `std`, so we'll have to generate mappings for the titlecase transforms.

One way we might do that is to roll some Ruby scripts like `focaccia` does for generating case folding mappings:

- https://github.com/artichoke/focaccia/blob/ed00eb45e6d1a38dda145cfae2b64c215119a2fe/scripts/gen_case_lookups.rb
- https://github.com/artichoke/focaccia/blob/ed00eb45e6d1a38dda145cfae2b64c215119a2fe/Rakefile#L92-L104

Another way we might tackle this is with codegen from crates in the Rust ecosystem:

- https://crates.io/crates/ucd-generate

Use the source layout of the `Lowercase` iterator as a guide:

- https://github.com/artichoke/roe/blob/acdab1e6b2b1eea475e8eea8a0a7c616ce7bf850/src/lowercase.rs
- https://github.com/artichoke/roe/tree/acdab1e6b2b1eea475e8eea8a0a7c616ce7bf850/src/lowercase
- https://github.com/artichoke/roe/blob/acdab1e6b2b1eea475e8eea8a0a7c616ce7bf850/src/lib.rs#L180-L320

## Implementation steps

- Implement `TitlecaseMode`.
- Implement ASCII iterator.
- Implement Full iterator.
- Wire everything up in a free function entrypoint in `lib.rs`.

### PR DESCRIPTION
**Pull request by choznerol**
Fix #65

> Unlike Lowercase and Uppercase, `char::to_titlecase` is not defined in `std`, so we'll have to generate mappings for the titlecase transforms.
>
> One way we might do that is to roll some Ruby scripts like focaccia does for generating case folding mappings:
>
> - https://github.com/artichoke/focaccia/blob/ed00eb45e6d1a38dda145cfae2b64c215119a2fe/scripts/gen_case_lookups.rb
> - https://github.com/artichoke/focaccia/blob/ed00eb45e6d1a38dda145cfae2b64c215119a2fe/Rakefile#L92-L104
>
> Another way we might tackle this is with codegen from crates in the Rust ecosystem:
>
> - https://crates.io/crates/ucd-generate

The `char::to_titlecase` was removed from `std` in https://github.com/rust-lang/rust/issues/26555.
Before actually start implementing either of the approach using `UnicodeData.txt`, I found a crate [`unicode_titlecase`] implementing `char::to_titlecase()` was created six months after #65 was opened.
If using this crate directly matches the project goal, the Titlecase implementation would be pretty similar to Uppercase and Lowercase (current approach); Otherwise, we'll have to do some mapping with `UnicodeData.txt` and `SpecialCasing.txt` similar to [how the create does](https://github.com/Teh-Bobo/unicode-title-case/blob/e7955f0157480621445a029920748b91ded388d4/build.rs#L12-L13).

[`unicode_titlecase`]: https://docs.rs/unicode_titlecase/2.0.0/unicode_titlecase/#traits

### CODE REVIEW

File: src/titlecase/full.rs
Comment (choznerol) [inline src/titlecase/full.rs]: I did a quick check with Ruby `String#capitalize`:

```shell
ruby -e 'puts ["ß", "ǅ", "İ", "İ", "ΈΤΟΣ", "ZȺȾ"].map{|c| "#{c} -> #{c.capitalize}\n#{c.codepoints} -> #{c.capitalize.codepoints}\n\n"}' 

ß -> Ss
[223] -> [83, 115]

ǅ -> ǅ
[453] -> [453]

İ -> İ
[304] -> [304]

İ -> İ
[304] -> [304]

ΈΤΟΣ -> Έτοσ
[904, 932, 927, 931] -> [904, 964, 959, 963]

ZȺȾ -> Zⱥⱦ
[90, 570, 574] -> [90, 11365, 11366]
```

File: src/titlecase/full.rs
Comment (choznerol) [inline src/titlecase/full.rs]: I copied and modified `case_map_to_(two|three)_chars` mainly from `tests::uppercase::full::tests`. The expected result matches with Ruby's `String#capitalize` returns. However, I'm not sure how those test cases were chosen initially, and if more/fewer cases are suitable for titlecase. For example, should I just remove the `ᾂ` test case considering it doesn't map to map to two chars here?

```shell
$  ruby -e 'puts ["case", "և", "ẙ", "ᾂ", "ﬗ", "ﬃ", "ὖ", "ῷ"].map{|c| "#{c} -> #{c.upcase} / #{c.capitalize} \n #{c.codepoints} -> #{c.upcase.codepoint
s} / #{c.capitalize.codepoints}\n\n"}'
case -> CASE / Case 
 [99, 97, 115, 101] -> [67, 65, 83, 69] / [67, 97, 115, 101]

և -> ԵՒ / Եւ 
 [1415] -> [1333, 1362] / [1333, 1410]

ẙ -> Y̊ / Y̊ 
 [7833] -> [89, 778] / [89, 778]

ᾂ -> ἊΙ / ᾊ 
 [8066] -> [7946, 921] / [8074]

ﬗ -> ՄԽ / Մխ 
 [64279] -> [1348, 1341] / [1348, 1389]

ﬃ -> FFI / Ffi 
 [64259] -> [70, 70, 73] / [70, 102, 105]

ὖ -> Υ̓͂ / Υ̓͂ 
 [8022] -> [933, 787, 834] / [933, 787, 834]

ῷ -> Ω͂Ι / ῼ͂ 
 [8183] -> [937, 834, 921] / [937, 834, 837]
```

**Comment by lopopolo** [comment]
> If using this crate directly matches the project goal

This crate seems to be ~unused in the ecosystem from an author I don't recognize. The repo itself has very few stars. I don't think I'd be comfortable taking a dependency on this crate if we don't need to. Can you explore the mapping approach instead?

File: Cargo.toml
Comment (lopopolo) [inline Cargo.toml]: I'm not comfortable using this dep

**Review by lopopolo — CHANGES_REQUESTED** [review]

**Comment by choznerol** [comment]
> > If using this crate directly matches the project goal
> 
> This crate seems to be ~unused in the ecosystem from an author I don't recognize. The repo itself has very few stars. I don't think I'd be comfortable taking a dependency on this crate if we don't need to. Can you explore the mapping approach instead?

Sure, let me explore the mapping approach 👍

File: Rakefile
Comment (choznerol) [inline Rakefile]: We only need `TITLE`, and it's surprisingly hard to ignore or autocorrect the unused `UPPER` and `LOWER` with the `dead_code` rule.

I'm trying to make it possible to generate only `TITLE` with a new `ucd-generate` flag  in https://github.com/BurntSushi/ucd-generate/pull/55.

```suggestion
    if system("ucd-generate case-mapping #{ucd_dir} --flat-table --include TITLE > #{dest}")
```

File: src/unicode/std_case_mapping_iter.rs
Comment (choznerol) [inline src/unicode/std_case_mapping_iter.rs]: Unfortunately [`std::char::CaseMappingIter`](https://doc.rust-lang.org/src/core/char/mod.rs.html#445) was not export, so I end up extracting part of them in this file.

File: Rakefile
Comment (b-n) [inline Rakefile]: I'm wondering if something like this to be a bit more specific to the changes, however I see the later cargo clippy, and couldn't find a way to make that run specifically in the directory, so maybe a full dirty check is just better. Just leaving the suggestion in case other things come to mind.

```suggestion
    unless system('git diff --exit-code -- "generated/**/*"')
```

File: Rakefile
Comment (b-n) [inline Rakefile]:
```suggestion
  ucd_dir = Pathname.pwd.join('generated', 'ucd')
```

likely will need `require 'pathname'` at the top too

File: Rakefile
Comment (b-n) [inline Rakefile]:
```suggestion
    filename = Pathname.pwd.join('generated', 'case_mapping.rs')
```

File: Rakefile
Comment (b-n) [inline Rakefile]:
```suggestion
        IO.copy_stream(data, "#{Pathname.pwd.join(ucd_dir, filename)}")
```

File: generated/ucd/README.md
Comment (b-n) [inline generated/ucd/README.md]:
```suggestion
Do not edit `generated/ucd/*.txt`. The files are automatically generated by:
```

File: generated/ucd/README.md
Comment (b-n) [inline generated/ucd/README.md:1]: I think this file should be in `generated/README.md` instead of this sub folder.

File: src/unicode.rs
Comment (b-n) [inline src/unicode.rs]:
```suggestion
#[cfg(feature = "alloc")]
pub use titlecase::{to_titlecase, Titlecase, ToTitlecase};
```

File: src/unicode.rs
Comment (b-n) [inline src/unicode.rs:1]: I'm wondering why the original structure didn't have `src/ascii.rs` in `src/ascii/mod.rs`. How to you feel about moving this to `src/unicode/mod.rs` and also moving `src/ascii.rs` to `src/ascii/mod.rs` ?

File: src/unicode/std_case_mapping_iter.rs
Comment (b-n) [inline src/unicode/std_case_mapping_iter.rs]:
```suggestion
// This file contains code extracted from std::char::CaseMappingIter.
```

File: src/titlecase.rs
Comment (b-n) [inline src/titlecase.rs]: Another example of the transformation into multiple codepoints I think would be really nice to see. e.g. the expansion of the ffi ligature and it's specific quirks since those are the crazy ones.

File: src/titlecase.rs
Comment (b-n) [inline src/titlecase.rs]: This file module is very nice! It lines up really nicely with the lower/uppercase modules!

File: src/uppercase/full.rs
Comment (b-n) [inline src/uppercase/full.rs:401]: :rocket:

File: src/lowercase/full.rs
Comment (b-n) [inline src/lowercase/full.rs:370]: :rocket:

File: src/titlecase/ascii.rs
Comment (b-n) [inline src/titlecase/ascii.rs:40]: This is absolutely genius way of doing this. I was thinking whether the self.first could have been made into a guard clause, but you'll always need the next char, so this is nice and efficient!

File: src/titlecase/ascii.rs
Comment (b-n) [inline src/titlecase/ascii.rs]: Thinking whether this should be checking the byte length, and validating that it shouldn't change rather than the output characters (as per the comment above it)

File: src/lib.rs
Comment (b-n) [inline src/lib.rs:614]: :heart:  new tests!

File: src/titlecase/full.rs
Comment (b-n) [inline src/titlecase/full.rs]: `first` instead of `beginning` to be consistent with the ascii implementation?

**Review by b-n — CHANGES_REQUESTED** [review]
A quick round of review on the "easy" parts :sweat_smile:. I'll come back and look at the more chunky `src/unicode/titlecase.rs` and `src/titlecase/full.rs` though, just wanted to give you some small nits quickly.

I should say, in general it looks really good!

File: Rakefile
Comment (choznerol) [inline Rakefile]: An older version of `ucd-generate` won't be caught by this. Let me actually check the installed version to be >= 0.3.0 too.

Comment (choznerol) [reply Rakefile]: Fixed in 2bbedfb.

Error for `ucd-generate` not installed:
> error ucd-generate not installed

Error for older version of `ucd-generate`:
> error please upgrade ucd-generate

File: src/titlecase/full.rs
Comment (choznerol) [reply src/titlecase/full.rs]: Thanks I didn't notice that one! Fixed in 01d0ac0.

File: src/titlecase/ascii.rs
Comment (choznerol) [reply src/titlecase/ascii.rs]: You're right. In fact, I realized `zȺȾ -> ZȺȾ` doesn't test much about the changing bytes length 😅. 
Fixed in 3e5fcee.

File: src/titlecase.rs
Comment (choznerol) [reply src/titlecase.rs]: Sure. Added in 2953a24.

File: src/unicode/std_case_mapping_iter.rs
Comment (choznerol) [reply src/unicode/std_case_mapping_iter.rs]: Fixed in 01d0ac0 too.

File: src/unicode.rs
Comment (choznerol) [reply src/unicode.rs:1]: I didn't find a difference between these two ways, but `src/ascii.rs -> src/ascii/mod.rs` also feel more intuitive to me too. Fix in 2fca461.

File: src/unicode.rs
Comment (choznerol) [reply src/unicode.rs]: But `Vec` isn't being used in `src/unicode/*` except in `tests`. I assume `cargo test --no-default-features` will catch this if `Vec` is actually used.
<details><summary>example</summary>
image
</details>

File: generated/ucd/README.md
Comment (choznerol) [reply generated/ucd/README.md:1]: Fixed in 01d0ac0 too.

File: generated/ucd/README.md
Comment (choznerol) [reply generated/ucd/README.md]: Thanks. Fixed in 01d0ac0.

File: Rakefile
Comment (choznerol) [reply Rakefile]: Fixed in 314f5b2315b660f588b7e12944bb01ae9fd90f19 too

File: Rakefile
Comment (choznerol) [reply Rakefile]: Fixed in 314f5b2315b660f588b7e12944bb01ae9fd90f19 too

File: Rakefile
Comment (choznerol) [reply Rakefile]: Thanks. Fixed in 314f5b2.
By the way, I end up using relative path in some places to avoid: image

File: Rakefile
Comment (choznerol) [reply Rakefile]: > couldn't find a way to make that run specifically in the directory

So do I. Seem like the only way is to have a separate package (`cargo clippy -p ...`).

The motivation for this check was to prevent the later `cargo clippy --fix` from modifying unstaged work unexpectedly. In contrast, I'm not worried too much about modifying `generated/**/*`, since they can be easily re-generated.

File: src/titlecase/ascii.rs
Comment (choznerol) [reply src/titlecase/ascii.rs:40]: I recall this one was ChatGPT's idea 😆

Comment (b-n) [reply src/titlecase/ascii.rs:40]: Ahh, the AI is here to take our jobs :sweat_smile:. You win this time ChatGPT!

File: src/unicode.rs
Comment (b-n) [reply src/unicode.rs]: Oh wow, good call! I did the naive thing of checking the ascii impl, and saw `to_ascii_lowercase` had the feature alloc around it.

The good news, after changing around the module structure, I now see what's going on I think.

It looks like `roe` currently exports not only the titlecase/lowercase/uppercase iterators, but also encoding specific functions - specifically things like `ascii::lowercase::to_ascii_lowercase`. I assume these functions are to be consumed by other Ruby string functions as short circuits for perf reasons (instead of slice => iterator => vec), and keeping all "case folding logic" in one crate.

This makes me wonder though, should `unicode::*` be exported or not. If yes, I think it would make sense to make `unicode::titlecase::to_titlecase` to be consistent with `ascii::titlecase::to_titlecase`, and thus to take a slice, and return a newly allocated `Vec<u8>`.

My preference:
- If we export `ascii::*` related methods, then we should also export `unicode::*` related things too
- `ascii::*` and `unicode::*` should take the same sort of parameters in this case

On "whether we should export `ascii::*`", I think it makes sense, and thus use `roe::ascii::*` instead of `std::str::*` in `artichoke-backend` and friends. @lopopolo thoughts?

Comment (b-n) [reply src/unicode.rs]: Note: if we do want to export `unicode::*`, then I would say for this PR we just don't export `unicode::*` for now, and then in the next PR do the work to bring some of the unicode functions into the `unicode::*` module etc.

File: src/titlecase/full.rs
Comment (b-n) [inline src/titlecase/full.rs]:
```suggestion
                    .expect("ToTitlecase or ToLowercase yields at least one char");
```

File: src/titlecase/full.rs
Comment (b-n) [reply src/titlecase/full.rs]: IMO, I would move the test cases that don't look right to the correct test case. e.g. moving:

```
ᾂ -> ἊΙ / ᾊ [8066] -> [7946, 921] / [8074]
```

to the `case_map_to_two_chars` based on what you've got here.

I suspect there might be some unicode [de]normalization going on here since U+1F82 can be denormalized into 4 codepoints: https://www.compart.com/en/unicode/U+1F82, and thus I suspect there is some other combinations around for this.

File: src/unicode/std_case_mapping_iter.rs
Comment (b-n) [inline src/unicode/std_case_mapping_iter.rs:1]: I hadn't checked through the need for this until just now, and I think I understand! Very cool.

So, to correct my understanding here, we need this since the iterator of a unicode titlecase because when we case map, we could end up with more unicode code points than what came in (e.g. 1 code point uppercase might be 3 lowercase chars). I'm trying to understand by reading the code through, but my brain is mush, and maybe it's easier for you to explain it to me - but my question is: Is it possible to wrap `core::char::ToLowerCase` and `core::char::ToUpperCase` iterators instead of bringing this in? It's entirely possible I'm missing something important here :sweat_smile:

**Review by b-n — DISMISSED** [review]
Found another tiny typo, but TBH it's looking really close to being there. I finally dived in and looked at the difference between `titlecase::full` and `lowercase::full`, and I see what you've done - very cool! I don't have many comments because of how close we are I think :smile:

File: src/unicode/std_case_mapping_iter.rs
Comment (choznerol) [reply src/unicode/std_case_mapping_iter.rs:1]: > we could end up with more unicode code points than what came in (e.g. 1 code point uppercase might be 3 lowercase chars).

Exactly 💯 at least that's also how I understand it!

> Is it possible to wrap `core::char::ToLowerCase` and `core::char::ToUpperCase` 

That's actually a very good idea! However, I gave it a try in https://github.com/artichoke/roe/commit/ada819cc8b4efca9e1af66b5c7da9875c75374f2#diff-1beb18474b7675a5d085ca3787c8cd13badcc4660db8aa63f6febb96ec71faa0L74-R84 but realized there wasn't a public API to construct a `ToLowercase`.

It appears [`char::to_lowercase()`](https://doc.rust-lang.org/stable/std/primitive.char.html#method.to_lowercase) is the only way to obtain a `ToLowercase`.  There isn't a method to mutate the internal of a [`ToLowercase`](https://doc.rust-lang.org/stable/std/char/struct.ToLowercase.html) either 😞.

Thanks for the idea still, and please let me know if you can think of any way that I can try. I'd really prefer to not copy `CaseMappingIter` if possible 😂.

File: src/titlecase/full.rs
Comment (choznerol) [reply src/titlecase/full.rs]: Fixed in 43f13a7a6332fce45e86b348a94f1d45f439a57b. Tks!

File: src/titlecase/full.rs
Comment (choznerol) [reply src/titlecase/full.rs]: > I would move the test cases that don't look right to the correct test case.

Ah, that make sense. Fixed in 09e6fe5df25cdd0d37b37bf7c8396d7a9487bcbf.

> I suspect there might be some unicode [de]normalization going on here since U+1F82 can be denormalized into 4 codepoints: https://www.compart.com/en/unicode/U+1F82, and thus I suspect there is some other combinations around for this.

I'm not sure if Unicode composition is related to case mapping 🤔 IIUC titlecasing is just a simple one directional mapping similar to uppercasing and lowercasing. 

https://www.compart.com/en/unicode/U+1F82 only shows its `Uppercase Character`, but https://codepoints.net/U+1F82 does show all its three kinds of case mapping.

**Review by choznerol — COMMENTED** [review]
Thanks for the detailed review!

File: src/titlecase/ascii.rs
Comment (lopopolo) [inline src/titlecase/ascii.rs:64]: There is a bug here with this sequence:

1. Create iterator
2. Call iter.next() to consume the title case / first case.
3. Call iter.next_back() until the iterator is consumed.

This will result in 2 uppercase ASCII chars being emitted.

File: src/unicode/std_case_mapping_iter.rs
Comment (lopopolo) [reply src/unicode/std_case_mapping_iter.rs:1]: Title case is a distinct operation from upper and lowercase.

This character is the example I know:

> Dž (titlecase form; all-capitals form DŽ, lowercase dž)

File: src/titlecase/ascii.rs
Comment (choznerol) [reply src/titlecase/ascii.rs:64]: Thanks! Reproduced in 42e1f79 and fixed in 6ffc252.

**Review by lopopolo — APPROVED** [review]
Added a couple of fixups. this looks great! Sorry this took so long to get approved.

Once this is merged, I'll prep a new release.

**Comment by lopopolo** [comment]
@b-n it looks like your :x: is only for a typo that @choznerol fixed. I'm gonna dismiss your review and get this merged

**Comment by b-n** [comment]
@lopopolo Ah yes, all good! I should have cleared my review 😅

**Comment by choznerol** [comment]
> Added a couple of fixups. this looks great! Sorry this took so long to get approved.
> 
> Once this is merged, I'll prep a new release.

No worry. Thank you guys for the review!

**Pull request merged**
